core/src/main/scala/org/apache/spark/shuffle/MemoizingShuffleDataIO.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.shuffle.api.{ShuffleDataIO, ShuffleDriverComponents, ShuffleExecutorComponents}
+
+/**
+ * Thin wrapper around {@link ShuffleDataIO} that ensures the given components are
+ * only initialized once and providing the same instance each time.
+ * <p>
+ * Used to ensure the SparkEnv only instantiates the given components once lazily
+ * and then reuses them throughout the lifetime of the SparkEnv.
+ */
+class MemoizingShuffleDataIO(delegate: ShuffleDataIO) {
+  private var _driver: ShuffleDriverComponents = _
+  private var _executor: ShuffleExecutorComponents = _
+
+  def getOrCreateDriverComponents(): ShuffleDriverComponents = synchronized {
+    if (_driver == null) {
+      _driver = delegate.initializeShuffleDriverComponents()
+    }
+    _driver
+  }
+
+  def getOrCreateExecutorComponents(
+      appId: String,
+      execId: String,
+      extraConfigs: Map[String, String]): ShuffleExecutorComponents = synchronized {
+    if (_executor == null) {
+      _executor = delegate.initializeShuffleExecutorComponents(appId, execId, extraConfigs.asJava)
+    }
+    _executor
+  }
```

Comment:
I think we can improve the methods of this class. Especially with `getOrCreateExecutorComponents` as reading that I have a feeling it is easy to use it badly by calling with two different values for any of its params right after each other as always the object generated for the first call will be returned.

But first `getOrCreateDriverComponents` that is the easier as here we have no input params at all so this could be replaced with a simple `lazy val` (it is thread-safe for a long time ago: https://github.com/scala/bug/issues/3007).

And the same is true for `getOrCreateExecutorComponents ` as all its parameters are basically coming from a SparkEnv instance.

So my idea is:
```suggestion

  lazy val driverComponents = delegate.initializeShuffleDriverComponents()

  lazy val executorComponents = {
    val env = SparkEnv.get
    delegate.initializeShuffleExecutorComponents(
      env.conf.getAppId,
      env.executorId,
      env.conf.getAllWithPrefix(ShuffleDataIOUtils.SHUFFLE_SPARK_CONF_PREFIX).toMap.asJava)
  }
```
I still have to test it. What is your opinion @mccheah ?

Reply:
I certainly gave `lazy val` a consideration. I'm not entirely familiar with the most modern Scala conventions - if `lazy` is preferred over explicit initialization methods in general, then I'm ok with the above recommendation.

Reply:
I think transforming `getOrCreateExecutorComponents` into method without any argument is already big win.
I can let go the `lazy val` it is just implementation details.

Reply:
LGTM, I'll make those changes. Thanks!